akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -261,32 +272,68 @@ private[akka] final class SeqStage[T] extends GraphStageWithMaterializedValue[Si
 
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
     val p: Promise[immutable.Seq[T]] = Promise()
-    val logic = new GraphStageLogic(shape) {
+    (new GraphStageLogic(shape) with InHandler {
+      setHandler(in, this)
+
       val buf = Vector.newBuilder[T]
 
       override def preStart(): Unit = pull(in)
 
-      setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        buf += grab(in)
+        pull(in)
+      }
 
-        override def onPush(): Unit = {
-          buf += grab(in)
-          pull(in)
-        }
+      override def onUpstreamFinish(): Unit = {
+        val result = buf.result()
+        p.trySuccess(result)
+        completeStage()
+      }
 
-        override def onUpstreamFinish(): Unit = {
-          val result = buf.result()
-          p.trySuccess(result)
-          completeStage()
-        }
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        p.tryFailure(ex)
+        failStage(ex)
+      }
+    }, p.future)
+  }
+}
 
-        override def onUpstreamFailure(ex: Throwable): Unit = {
-          p.tryFailure(ex)
-          failStage(ex)
-        }
-      })
-    }
+/**
+ * INTERNAL API
+ */
+private[akka] final class CountStage[T] extends GraphStageWithMaterializedValue[SinkShape[T], Future[IOResult]] {
+  val in = Inlet[T]("count.in")
 
-    (logic, p.future)
+  override def toString: String = "CountStage"
+
+  override val shape: SinkShape[T] = SinkShape.of(in)
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.countSink
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    val p: Promise[IOResult] = Promise()
+    (new GraphStageLogic(shape) with InHandler {
+      setHandler(in, this)
```

Comment:
probably doesn't matter, but feels better to do this as the last thing in the constructor (since it's "leaking" this)